File: Python/Lecture_04_Lab_Exercise_1_Undamped.py
'''
========================================================================
        ╦ ╦┌─┐┌─┐┌─┐┌─┐┌┬┐  ╔╦╗┌─┐┌─┐┌┬┐┬ ┬  ╔╗ ┌─┐┬  ┌─┐┬ ┬┌─┐
        ╠═╣│ │└─┐└─┐├─┤│││  ║║║├─┤│ ┬ ││└┬┘  ╠╩╗├─┤│  ├─┤├─┤├─┤
        ╩ ╩└─┘└─┘└─┘┴ ┴┴ ┴  ╩ ╩┴ ┴└─┘─┴┘ ┴   ╚═╝┴ ┴┴─┘┴ ┴┴ ┴┴ ┴
========================================================================
# Author: Hossam Magdy Balaha
# Permissions and Citation: Refer to the README file.
'''

# Import necessary libraries.
import numpy as np
import matplotlib.pyplot as plt

# Physical parameters: natural frequency and amplitude used to build x(t).
omega0 = 2 * np.pi  # Set the natural frequency (1 Hz, in radians per second).
A = 1.0  # Set the amplitude of oscillation.

# Generate a high-resolution time vector for smooth visualization of oscillations.
t = np.linspace(0, 5, 1000)

# Compute displacement for two different phase/amplitude choices (B).
# Case 1: no phase shift (B = 0.0).
B = 0.0  # Set the second amplitude to zero (no phase shift).
x1 = A * np.cos(omega0 * t) + B * np.sin(omega0 * t)

# Case 2: with phase contribution (B = 1.0).
# Changing B alters the initial condition and shifts the waveform.
B = 1.0  # Set the second amplitude to one (phase shift).
x2 = A * np.cos(omega0 * t) + B * np.sin(omega0 * t)

# Plot the two waveforms to compare the effect of the phase term.
plt.figure(figsize=(10, 5))  # Create a figure with a specified size.
plt.plot(t, x1, label="B = 0.0", color="blue")  # Plot the first displacement with label and color.
plt.plot(t, x2, label="B = 1.0", color="orange")  # Plot the second displacement with label and color.
plt.title("Undamped Oscillation")  # Set the plot title.
plt.xlabel("Time (t)")  # Label the x-axis as time.
plt.ylabel("Displacement x(t)")  # Label the y-axis as displacement.
plt.grid(True)  # Enable the grid for better readability.
plt.legend()  # Show the legend on the plot.
plt.tight_layout()  # Adjust the layout to prevent overlap of labels and titles.

# Save the plot as a PNG file with high resolution for inclusion in lecture notes.
plt.savefig("Lecture_04_Lab_Exercise_1_Undamped.png", dpi=300, bbox_inches="tight")

# Display the plot interactively.
plt.show()

File: MATLAB/Lecture_04_Lab_Exercise_1_Undamped.m
% ===========================================================================
%         ╦ ╦┌─┐┌─┐┌─┐┌─┐┌┬┐  ╔╦╗┌─┐┌─┐┌┬┐┬ ┬  ╔╗ ┌─┐┬  ┌─┐┬ ┬┌─┐
%         ╠═╣│ │└─┐└─┐├─┤│││  ║║║├─┤│ ┬ ││└┬┘  ╠╩╗├─┤│  ├─┤├─┤├─┤
%         ╩ ╩└─┘└─┘└─┘┴ ┴┴ ┴  ╩ ╩┴ ┴└─┘─┴┘ ┴   ╚═╝┴ ┴┴─┘┴ ┴┴ ┴┴ ┴
% ===========================================================================
%
% Author: Hossam Magdy Balaha
% Permissions and Citation: Refer to the README file.

% Set the parameters.
omega0 = 2 * pi; % Natural frequency (1 Hz, radians per second).
A = 1.0; % Amplitude of oscillation.

% Generate 1000 time points for high-resolution plotting.
t = linspace(0, 5, 1000);

% Calculate displacement x(t) for two cases of phase/amplitude B.
% Case 1: B = 0.0 (no phase shift).
B1 = 0.0;
x1 = A * cos(omega0 * t) + B1 * sin(omega0 * t);

% Case 2: B = 1.0 (with phase shift).
B2 = 1.0;
x2 = A * cos(omega0 * t) + B2 * sin(omega0 * t);

% Plotting.
figure;  % Create a new figure.
plot(t, x1, 'b', 'LineWidth', 1.5); 
hold on;
plot(t, x2, 'r--', 'LineWidth', 1.5);

title('Undamped Oscillation');
xlabel('Time (t)');
ylabel('Displacement x(t)');
legend({'B = 0.0', 'B = 1.0'});
grid on;
axis tight; % Adjust axis limits to fit data tightly.
hold off;

% Save the plot as a PNG file for lecture materials.
saveas(gcf, 'Lecture_04_Lab_Exercise_1_Undamped.png');